### ISSUE TITLE
Better bit flag export

### ISSUE THREAD

**Issue by JimmyJames707** [godotengine/godot#9308]
So there's an option in GDScript to export bit flags, but it sets the values for you. To get the functionality I need, I have to make another dictionary.
```
enum MY_VAL {
    NONE                = 0x00,
    FLAG_1              = 0x11,
    FLAG_2              = 0x12,
    FLAG_3              = 0x14,
    FLAG_4              = 0x21,
    FLAG_5              = 0x22,
    FLAG_6              = 0x24,
    FLAG_7              = 0x41,
    FLAG_8              = 0x42,
    FLAG_9              = 0x44,
    MASK_1              = 0x01,
    MASK_2              = 0x02,
    MASK_3              = 0x03,
    MASK_4              = 0x10,
    MASK_5              = 0x20,
    MASK_6              = 0x30,
}

var MY_VAL_TO_HEX = {
    0 : MY_VAL.NONE,
    1 : MY_VAL.FLAG_1,
    2 : MY_VAL.FLAG_2,
    3 : MY_VAL.FLAG_3,
    4 : MY_VAL.FLAG_4,
    5 : MY_VAL.FLAG_5,
    6 : MY_VAL.FLAG_6,
    7 : MY_VAL.FLAG_7,
    8 : MY_VAL.FLAG_8,
    9 : MY_VAL.FLAG_9,
    10 : MY_VAL.MASK_1,
    11 : MY_VAL.MASK_2,
    12 : MY_VAL.MASK_3,
    13 : MY_VAL.MASK_4,
    14 : MY_VAL.MASK_5,
    15 : MY_VAL.MASK_6,
}

export (int,                  \
    "NONE",                 \
    "FLAG_1",               \
    "FLAG_2",               \
    "FLAG_3",               \
    "FLAG_4",               \
    "FLAG_5",               \
    "FLAG_6",               \
    "FLAG_7",               \
    "FLAG_8",               \
    "FLAG_9",               \
    "MASK_1",               \
    "MASK_2",               \
    "MASK_3",               \
    "MASK_4",               \
    "MASK_5",               \
    "MASK_6")               \
var my_val = 0

func do_stuff():
    var correct_my_val = MY_VAL_TO_HEX[my_val]
```

My suggestion is to make it so the export can do something like below. 
```
export (int, FLAGS,        \
    {"NONE" = 0x00},                 \
    {"FLAG_1" = 0x11},               \
```

**Comment by bojidar-bg** [comment]
> My suggestion is to make it so the export can do something like below.
> ```gdscript
> export (int, FLAGS,        \
>     {"NONE" = 0x00},                 \
>     {"FLAG_1" = 0x11},               \
> ```
This won't work per-se as those aren't valid dictionaries 😃 . (remove the quotes around the keys)

Just to make sure -- you want to be able to supply two dictionaries, one with exact values and one with masks?

**Comment by JimmyJames707** [comment]
"you want to be able to supply two dictionaries, one with exact values and one with masks?"

No, I want to be able to export one enum with values of my choosing.

The current way to export bit flags is like so:

export (int, FLAGS,								\
	"NAME1",									\
	"NAME2",									\
        etc
var flags = 0

and the engine assigns the values for you in this order:

0x00000001,
0x00000002,
0x00000004,
0x00000008,
0x00000010,
0x00000020,
0x00000040,
0x00000080,
etc

What I want is be able to set the hex values that are to be exported., so I can have an enum like this, without using something like MY_VAL_TO_HEX in the OP to look up the values.

enum MY_VAL {
    NONE                = 0x00,
    FLAG_1              = 0x11,
    FLAG_2              = 0x12,
    FLAG_3              = 0x14,
    FLAG_4              = 0x21,
    FLAG_5              = 0x22,
    FLAG_6              = 0x24,
    FLAG_7              = 0x41,
    FLAG_8              = 0x42,
    FLAG_9              = 0x44,
    MASK_1              = 0x01,
    MASK_2              = 0x02,
    MASK_3              = 0x03,
    MASK_4              = 0x10,
    MASK_5              = 0x20,
    MASK_6              = 0x30,
}

**Comment by JimmyJames707** [comment]
Another idea I had is give the export the name of the enum and it gets the hex values from the enum.

```
enum MY_VAL {
    NONE                = 0x00,
    FLAG_1              = 0x11,
}
```

So something like this:

```
export (MY_VAL)
var flags = 0
```

**Comment by bojidar-bg** [comment]
Ok, sounds legit... 😃 Actually this was proposed way back on the enum PR, but never implemented so far.

**Comment by JimmyJames707** [comment]
Just thought I'd add that my first idea while it is longer to type, allows one to use constants as well as enums. 
```
export (int, FLAGS,        \
    {"NONE" = 0x00},                 \
    {"FLAG_1" = 0x11},               \
```

**Comment by kubecz3k** [comment]
First of all thank you for your report and sorry for the delay.

We released Godot 3.0 in January 2018 after 18 months of work, fixing many old issues either directly, or by obsoleting/replacing the features they were referring to.

We still have hundreds of issues whose relevance/reproducibility needs to be checked against the [current stable version](https://godotengine.org/download), and that's where you can help us.
**Could you check if the issue that you described initially is still relevant/reproducible in Godot 3.0 or any newer version, and comment about its current status here?**

For bug reports, please also make sure that the issue contains detailed steps to reproduce the bug and, if possible, a zipped project that can be used to reproduce it right away. This greatly speeds up debugging and bugfixing tasks for our contributors.

Our Bugsquad will review this issue more in-depth in 15 days, and potentially close it if its relevance could not be confirmed.

Thanks in advance.

*Note: This message is being copy-pasted to many "stale" issues (90+ days without activity). It might happen that it is not meaningful for this specific issue or appears oblivious of the issue's context, if so please comment to notify the Bugsquad about it.*

**Comment by bojidar-bg** [comment]
This is actually now resolved as #12952 was merged.
It even sort-of works with flags, but, YMMV.
